File: sdk/typescript/src/tunnels/client/_callable_streaming.ts
/**
 * inkbox-tunnels/client/_callable_streaming.ts
 *
 * Streaming wrapper for `InkboxHandler` used by `CallableDispatch` in
 * passthrough mode. The Fetch-style handler returns a `Response`; this
 * module pipes that response back through `DispatchResponseSink` so
 * the third party gets a true streamed reply.
 *
 * WebSocket upgrades are handled separately by
 * `_ws_passthrough.bridgeWsHandlerOverSink` — the h1 parser and h2
 * transcoder route `isWebSocket=true` requests directly to
 * `CallableDispatch.dispatchWebSocket` so they never reach this
 * HTTP-only invoker.
 */

import type {
  DispatchRequest,
  DispatchResponseSink,
} from "./_dispatch.js";
import type { InkboxHandler, InkboxRequestContext } from "./_handler.js";
import type { ReadonlyEnvelope } from "./_envelope.js";
import { HOP_BY_HOP_RESPONSE } from "./_protocol.js";

export interface InvokeHandlerOpts {
  handler: InkboxHandler;
  request: DispatchRequest;
  response: DispatchResponseSink;
  publicHost: string;
  maxOutboundBodyBytes: number;
}

/**
 * Build a minimal envelope-shaped object for the InkboxRequestContext.
 * Passthrough callable dispatch doesn't have a true envelope (those
 * are produced by the tunnel server's intake path), so we synthesize a
 * placeholder. The handler can read the typed fields directly via
 * `req` / `ctx` instead of poking at this for normal use.
 */
function synthesizeEnvelope(req: DispatchRequest): ReadonlyEnvelope {
  return Object.freeze({
    requestId: "",
    method: req.method,
    path: req.path,
    routeKind: "webhook" as const,
    wsId: null,
    forwardedHeaders: req.headers,
    body: Buffer.alloc(0),
    bodyUri: null,
    forwardedForIp: req.forwardedForIp,
    tcpId: null,
    sniHost: req.sniHost,
    extraMeta: {},
  });
}

async function bodyToReadable(
  iter: AsyncIterable<Buffer>,
): Promise<ReadableStream<Uint8Array> | null> {
  const it = iter[Symbol.asyncIterator]();
  let probed = await it.next();
  if (probed.done) return null;
  const first = probed.value;
  return new ReadableStream<Uint8Array>({
    async start(controller) {
      controller.enqueue(new Uint8Array(first));
      try {
        while (true) {
          const r = await it.next();
          if (r.done) break;
          controller.enqueue(new Uint8Array(r.value));
        }
        controller.close();
      } catch (err) {
        controller.error(err);
      }
    },
  });
}

export async function invokeHandlerStreaming(
  opts: InvokeHandlerOpts,
): Promise<void> {
  const { handler, request, response, publicHost, maxOutboundBodyBytes } =
    opts;

  // Build a Fetch Request from the DispatchRequest. Headers are flat,
  // method/path provided. URL host is the public host.
  const headers = new Headers();
  for (const [k, v] of request.headers) {
    if (k.startsWith(":")) continue;
    try {
      headers.append(k, v);
    } catch {
      /* invalid header value — skip */
    }
  }
  if (!headers.has("host")) headers.set("host", publicHost);
  if (!headers.has("x-forwarded-host")) headers.set("x-forwarded-host", publicHost);
  if (!headers.has("x-forwarded-proto")) headers.set("x-forwarded-proto", "https");
  if (request.forwardedForIp != null) {
    if (!headers.has("x-forwarded-for"))
      headers.set("x-forwarded-for", request.forwardedForIp);
  }

  const url = `https://${publicHost}${request.path.startsWith("/") ? "" : "/"}${request.path}`;

  const bodyStream =
    request.method === "GET" || request.method === "HEAD"
      ? null
      : await bodyToReadable(request.body);

  const fetchReq = new Request(url, {
    method: request.method,
    headers,
    body: bodyStream as BodyInit | null,
    // Required when sending a stream body on Node fetch.
    duplex: "half",
  } as RequestInit & { duplex?: "half" });

  const ctx: InkboxRequestContext = {
    signal: new AbortController().signal,
    forwardedForIp: request.forwardedForIp,
    sniHost: request.sniHost,
    envelope: synthesizeEnvelope(request),
  };

  let resp: Response;
  try {
    resp = await handler(fetchReq, ctx);
  } catch {
    await response.sendHead({
      status: 502,
      headers: [["content-type", "text/plain"]],
    });
    await response.sendBody(Buffer.from("upstream error"));
    await response.endBody();
    return;
  }

  const respHeaders: Array<[string, string]> = [];
  resp.headers.forEach((value, key) => {
    if (HOP_BY_HOP_RESPONSE.has(key.toLowerCase())) return;
    respHeaders.push([key, value]);
  });

  await response.sendHead({ status: resp.status, headers: respHeaders });
  if (resp.body == null) {
    await response.endBody();
    return;
  }

  let bytesOut = 0;
  const reader = resp.body.getReader();
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      const buf = Buffer.from(value);
      bytesOut += buf.length;
      if (bytesOut > maxOutboundBodyBytes) {
        await response.reset("response-too-large");
        return;
      }
      await response.sendBody(buf);
    }
    await response.endBody();
  } catch {
    await response.reset("upstream-error");
  } finally {
    try {
      reader.releaseLock();
    } catch {
      /* swallow */
    }
  }
}

File: sdk/python/inkbox/tunnels/client/_callable_streaming.py
"""
inkbox/tunnels/client/_callable_streaming.py

Streaming ASGI HTTP invoker used by ``CallableDispatch`` in passthrough
mode. Unlike the buffered ``invoke_asgi_http`` in ``_asgi.py`` (which is
fine for the edge envelope path that materializes the whole response),
this module drives ASGI with a streaming ``receive`` (yielding inbound
body chunks as they arrive) and a streaming ``send`` (translating
``http.response.body`` events into ``DispatchResponseSink.send_body``
calls so the third party gets a true streamed response).

WebSocket upgrades are handled separately by
``_ws_passthrough.invoke_asgi_websocket`` — the h1 parser and h2
transcoder route ``is_websocket=True`` requests directly to
``CallableDispatch.dispatch_websocket`` so they never reach this
HTTP-only invoker.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any

from inkbox.tunnels.client._asgi import build_asgi_http_scope
from inkbox.tunnels.client._dispatch import (
    DispatchRequest,
    DispatchResponseHead,
    DispatchResponseSink,
)


logger = logging.getLogger("inkbox.tunnels")


async def invoke_asgi_streaming(
    app: Any,
    request: DispatchRequest,
    response: DispatchResponseSink,
    *,
    public_host: str,
    max_outbound_body_bytes: int,
) -> None:
    """Drive ``app`` (an ASGI HTTP app) against ``request``; stream the
    response back through ``response``."""
    scope = build_asgi_http_scope(
        method=request.method,
        path=request.path,
        headers=request.headers,
        forwarded_for_ip=request.forwarded_for_ip,
        public_host=public_host,
    )

    body_iter = request.body
    body_done = asyncio.Event()
    body_buffer: list[bytes] = []
    fetch_lock = asyncio.Lock()

    async def _next_chunk() -> tuple[bytes, bool]:
        async with fetch_lock:
            if body_buffer:
                return body_buffer.pop(0), False
            if body_done.is_set():
                return b"", True
            try:
                chunk = await body_iter.__anext__()
                return chunk, False
            except StopAsyncIteration:
                body_done.set()
                return b"", True

    async def receive() -> dict[str, Any]:
        chunk, done = await _next_chunk()
        return {
            "type": "http.request",
            "body": chunk,
            "more_body": not done,
        }

    head_sent = False
    bytes_out = 0
    aborted = False

    async def send(event: dict[str, Any]) -> None:
        nonlocal head_sent, bytes_out, aborted
        if aborted:
            return
        etype = event.get("type")
        if etype == "http.response.start":
            status = int(event.get("status", 200))
            raw_headers = event.get("headers", []) or []
            decoded: list[tuple[str, str]] = []
            for k, v in raw_headers:
                kk = (
                    k.decode("latin-1") if isinstance(k, (bytes, bytearray))
                    else str(k)
                )
                vv = (
                    v.decode("latin-1") if isinstance(v, (bytes, bytearray))
                    else str(v)
                )
                decoded.append((kk, vv))
            await response.send_head(
                DispatchResponseHead(status=status, headers=decoded),
            )
            head_sent = True
        elif etype == "http.response.body":
            body = event.get("body", b"") or b""
            if isinstance(body, str):
                body = body.encode("utf-8")
            if not head_sent:
                # ASGI server should always send start before body.
                await response.send_head(
                    DispatchResponseHead(
                        status=200,
                        headers=[("content-type", "application/octet-stream")],
                    ),
                )
                head_sent = True
            bytes_out += len(body)
            if bytes_out > max_outbound_body_bytes:
                aborted = True
                await response.reset("response-too-large")
                return
            if body:
                await response.send_body(body)
            if not event.get("more_body", False):
                await response.end_body()
        # 'http.disconnect' is for handler-side observation; we don't
        # synthesize it from the SDK side in this minimal v1.

    try:
        await app(scope, receive, send)
    except Exception:
        logger.exception("ASGI handler raised")
        if not head_sent:
            try:
                await response.send_head(
                    DispatchResponseHead(
                        status=502,
                        headers=[("content-type", "text/plain")],
                    ),
                )
                await response.send_body(b"upstream error")
                await response.end_body()
            except Exception:
                pass
